benchmark_data_loading/dataset_helpers.py

```diff
@@ -102,28 +102,31 @@ def _make_dp_from_tars(*, root, archive_size):
     return dp
 
 
-def _make_webdataset(*, root, archive_size):
-    archives = Path(root).glob(f"archive_{archive_size}*.tar")
-    archives = [str(a) for a in archives]
-    return wds.WebDataset(archives)  # This will read and load the data as bytes
-
-
 def make_dp(*, root, archive=None, archive_content=None, archive_size=500):
     if archive in ("pickle", "torch"):
         dp = _make_dp_from_archive(
             root=root, archive=archive, archive_content=archive_content, archive_size=archive_size
         )
     elif archive == "tar":
         dp = _make_dp_from_tars(root=root, archive_size=archive_size)
-    elif archive == "webdataset":
-        dp = _make_webdataset(root=root, archive_size=archive_size)
     else:
         dp = _make_dp_from_image_folder(root=root)
 
     dp = LenSetter(dp)
     return dp
 
 
+def make_webdataset(*, root, archive_size=500):
+    # Don't use this without `with_DL`, because the intra-archive shuffling only happens there
+    archives = Path(root).glob(f"archive_{archive_size}*.tar")
+    archives = [str(a) for a in archives]
+    return (
+        wds.WebDataset(archives)
+        .shuffle(len(archives), initial=len(archives))  # inter-archive shuffling
+        .map(lambda sample: io.BytesIO(sample["jpeg"]))
```

Comment:
For this to be more comparable with our own `.tar` and other archives, perhaps we should only read the file data here (i.e. use `_read_tar_entry` or `just_read_the_file` or something similar), and only convert to `BytesIO` later (outside of `make_dataset()`?

Reply:
This is just pulling an entry from a dictionary and wrapping it into a `io.BytesIO`. That is really fast even for Python standards:

```py
import io
from time import perf_counter_ns

import torch


with open("test/assets/encode_jpeg/grace_hopper_517x606.jpg", "rb") as file:
    jpeg = file.read()

sample = {"jpeg": jpeg}

fn = lambda sample: io.BytesIO(sample["jpeg"])

times = []
for _ in range(1_000_000):
    start = perf_counter_ns()
    fn(sample)
    stop = perf_counter_ns()
    times.append(stop - start)
times = torch.tensor(times, dtype=torch.float64)

print(f"median={float(torch.median(times)):.0f} ns, std={float(torch.std(times)):.0f} ns")
```

```
median=260 ns, std=37 ns
```

I don't think we should worry about sub microseconds. The other dataset creators also perform some "cleanup" like dropping the label, so I would keep it as is.